S3 config should not stomp local secret config

coreydaley opened the pull request:
Cluster S3 configuration should not stomp user configuration
    
  - Checks for a user supplied S3 configuration to use
  - If user config is not found, use the system configuration
  - Add a test to ensure that a user provided configuration will override the sytem provided one
  - Add image-registry-private-configuration-user to the README.md
  - Finalize S3 storage (remove bucket if managed by us)
  - Don't fail on S3 bucket tagging and encryption, just set the appropriate conditions to false

Fixes #79 
Fixes #99

coreydaley commented on pkg/clusterconfig/clusterconfig.go:
```diff
@@ -106,12 +113,21 @@ func GetAWSConfig() (*Config, error) {
+		return nil, err
 	}
 
-	cfg.Storage.Type = StorageTypeS3
```
Removing this since it is defined earlier in the function.

coreydaley commented:
/retest

coreydaley commented:
/retest

coreydaley commented:
/retest

bparees commented on pkg/clusterconfig/clusterconfig.go:
```diff
@@ -25,6 +30,8 @@ const (
 	installerConfigNamespace = "kube-system"
 	installerConfigName      = "cluster-config-v1"
 	installerAWSCredsName    = "aws-creds"
+	imageRegistryNamespace   = "openshift-image-registry"
```
we don't have this somewhere else already?

coreydaley replied:
updated in next push

bparees commented on pkg/clusterconfig/clusterconfig.go:
```diff
@@ -25,6 +30,8 @@ const (
 	installerConfigName      = "cluster-config-v1"
 	installerAWSCredsName    = "aws-creds"
+	imageRegistryNamespace   = "openshift-image-registry"
+	userPrivateConfigName    = "image-registry-private-configuration-user"
```
where's the `image-registry-private-configuration` constant defined?  can be put this with that or vice versa?

coreydaley replied:
It wasn't, but i added it in the next push

bparees commented on pkg/storage/s3/s3.go:
```diff
@@ -122,12 +122,12 @@ func (d *driver) createAndTagBucket(svc *s3.S3, installConfig *installer.Install
+	}
 
-	return util.CreateOrUpdateSecret("image-registry", "openshift-image-registry", data)
+	return util.CreateOrUpdateSecret("image-registry-private-configuration", "openshift-image-registry", data)
```
constant

coreydaley replied:
updated in next push

bparees commented on test/e2e/aws_test.go:
```diff
@@ -184,4 +185,22 @@ func TestAWS(t *testing.T) {
+	}
+	if fakeAWSCredsData["REGISTRY_STORAGE_S3_ACCESSKEY"] != cfgUser.Storage.S3.AccessKey || fakeAWSCredsData["REGISTRY_STORAGE_S3_SECRETKEY"] != cfgUser.Storage.S3.SecretKey {
+		t.Errorf("expected system configuration to be overridden by the user configuration but it wasn't.")
+	}
```
since you've gotten this far, you might as well add a test that confirms we report a useful status condition when you can't create the s3 bucket because the credentials are invalid.

coreydaley replied:
updated in current push

bparees commented on pkg/clusterconfig/clusterconfig.go:
```diff
@@ -102,21 +109,38 @@ func GetAWSConfig() (*Config, error) {
+	} else if err != nil {
+		return nil, err
+	} else {
+		glog.Infof("Using user defined AWS credentials from secret \"%s/%s\"", imageRegistryNamespace, userPrivateConfigName)
```
need to handle watching this secret in case the value is changed (or the secret is introduced later.  or deleted later).

coreydaley replied:
Should I introduce that code directly into https://github.com/openshift/cluster-image-registry-operator/blob/master/pkg/operator/controller.go#L162 which runs periodically , or should it be some kind of separate watcher?

dmage replied:
@coreydaley it handles events from watchers: https://github.com/openshift/cluster-image-registry-operator/blob/83f0b8d3eddf2175496f915b7c095baa1d7f57a9/pkg/operator/controller.go#L282-L292

bparees replied:
yeah it should be part of the existing controller which is already watching secrets anyway, you just need to add the logic to handle *this* secret.

dmage commented on test/e2e/aws_test.go:
```diff
@@ -184,4 +185,22 @@ func TestAWS(t *testing.T) {
+	}
+
+	if err := util.CreateOrUpdateSecret("image-registry-private-configuration-user", "openshift-image-registry", fakeAWSCredsData); err != nil {
+		t.Errorf("unable to create secret \"openshift-image-registry/image-registry-configuration-user\": %#v", err)
```
Why isn't it a fatal error?

dmage commented on pkg/clusterconfig/clusterconfig.go:
```diff
@@ -102,21 +109,38 @@ func GetAWSConfig() (*Config, error) {
+		return nil, err
+	} else {
+		glog.Infof("Using user defined AWS credentials from secret \"%s/%s\"", imageRegistryNamespace, userPrivateConfigName)
+		if v, ok := sec.Data["REGISTRY_STORAGE_S3_ACCESSKEY"]; ok {
```
What if the user has made a typo and the key isn't present in the secret?

coreydaley replied:
updated in next push

dmage commented on pkg/clusterconfig/clusterconfig.go:
```diff
@@ -102,21 +109,38 @@ func GetAWSConfig() (*Config, error) {
+		// If no user defined secret is found, use the system one
+		sec, err = client.Secrets(installerConfigNamespace).Get(installerAWSCredsName, metav1.GetOptions{})
+		if err != nil {
+			return nil, fmt.Errorf("unable to get secret  \"%s/%s\": %v", installerConfigNamespace, installerAWSCredsName, err)
```
two spaces, and it's better to use %q to get a quoted string

coreydaley replied:
updated in next push

dmage commented on pkg/clusterconfig/clusterconfig.go:
```diff
@@ -102,21 +109,38 @@ func GetAWSConfig() (*Config, error) {
+		if err != nil {
+			return nil, fmt.Errorf("unable to get secret  \"%s/%s\": %v", installerConfigNamespace, installerAWSCredsName, err)
+		}
+		glog.Infof("Using cluster defined AWS credentials from secret \"%s/%s\"", installerConfigNamespace, installerAWSCredsName)
```
I don't think we want to have two info messages on each event. If it's important to indicate which configuration is in use, it's better to use status.

coreydaley replied:
updated in next push

dmage commented on pkg/clusterconfig/clusterconfig.go:
```diff
@@ -102,21 +109,38 @@ func GetAWSConfig() (*Config, error) {
+			return nil, fmt.Errorf("unable to get secret  \"%s/%s\": %v", installerConfigNamespace, installerAWSCredsName, err)
+		}
+		glog.Infof("Using cluster defined AWS credentials from secret \"%s/%s\"", installerConfigNamespace, installerAWSCredsName)
+		if v, ok := sec.Data["aws_access_key_id"]; ok {
```
There should be a warning or an error if the key isn't in the secret.

coreydaley replied:
updated in next push

legionus commented on pkg/clusterconfig/clusterconfig.go:
```diff
@@ -102,21 +109,38 @@ func GetAWSConfig() (*Config, error) {
+		if v, ok := sec.Data["REGISTRY_STORAGE_S3_ACCESSKEY"]; ok {
+			cfg.Storage.S3.AccessKey = string(v)
+		}
+		if v, ok := sec.Data["REGISTRY_STORAGE_S3_SECRETKEY"]; ok {
```
Please check that the `image-registry-private-configuration-user` does not contain any unknown keys to avoid typos.

coreydaley replied:
updated in next push

legionus commented on pkg/apis/imageregistry/v1alpha1/types.go:
```diff
@@ -10,7 +10,10 @@ import (
 const (
-	OperatorStatusTypeRemoved = "Removed"
+	OperatorStatusTypeRemoved             = "Removed"
+	OperatorNamespace                     = "openshift-image-registry"
```
No. Use: 
https://github.com/openshift/cluster-image-registry-operator/blob/master/pkg/client/client.go#L65-L74

coreydaley replied:
Did you mean https://github.com/openshift/cluster-image-registry-operator/blob/master/pkg/client/client.go#L56 to get the namespace?

coreydaley replied:
Fixed in next push.

dmage commented on pkg/clusterconfig/clusterconfig.go:
```diff
@@ -116,30 +115,37 @@ func GetAWSConfig() (*Config, error) {
 		if v, ok := sec.Data["aws_access_key_id"]; ok {
 			cfg.Storage.S3.AccessKey = string(v)
+		} else {
+			return nil, fmt.Errorf("Secret %q does not contain required key \"aws_access_key_id\"", fmt.Sprintf("%s/%s", regopapi.OperatorNamespace, regopapi.ImageRegistryPrivateConfiguration))
```
sec is a secret from the installer.

bparees replied:
?

dmage replied:
The error mentions `fmt.Sprintf("%s/%s", regopapi.OperatorNamespace, regopapi.ImageRegistryPrivateConfiguration)`, but the lookup was in the installer secret.

bparees replied:
ah yes, good catch!

coreydaley replied:
updated in next push

coreydaley commented:
Blocked by https://github.com/openshift/cluster-image-registry-operator/issues/75

coreydaley commented:
@bparees ptal

bparees commented on pkg/resource/generator.go:
```diff
@@ -86,8 +96,132 @@ func (g *Generator) list(cr *regopapi.ImageRegistry) ([]Mutator, error) {
+	}
+
+	// Get the existing image-registry-private-configuration secret
+	sec, err := client.Secrets(g.params.Deployment.Namespace).Get(regopapi.ImageRegistryPrivateConfiguration, metav1.GetOptions{})
```
as a rule for content we are already watching, and thus have cached in a listwatcher, we should retrieve it from the cache, not make a new api call.  There are a few spots in this PR that this applies to.  I'm not going to make you fix it in this PR, but it's something to keep in mind and come back to.

coreydaley replied:
ok, i will keep that in mind in the future

bparees commented on pkg/resource/generator.go:
```diff
@@ -86,8 +96,132 @@ func (g *Generator) list(cr *regopapi.ImageRegistry) ([]Mutator, error) {
+					return err
+				}
+				// If the bucket doesn't exist, try to create it
+				if err := driver.CreateAndTagBucket(installConfig, cr); err != nil {
```
if we create the bucket (after a name change, or basically ever) that has implications for the ManagedStorage value, so it needs to be updated, right?

(Suppose we started out using a bucket they created, so ManagedStorage=false, we won't delete the bucket on teardown...  then they change the bucketname to something else and we *do* create the bucket...now we own it.  If they change the bucket name to something else that already exists, we don't own it.  So you need to be keeping that state updated)

coreydaley replied:
This should be updated now.

bparees commented on pkg/resource/generator.go:
```diff
@@ -86,8 +96,132 @@ func (g *Generator) list(cr *regopapi.ImageRegistry) ([]Mutator, error) {
+		} else if cr.Spec.Storage.S3.Bucket != "" {
+			// If the bucket name didn't change,
+			// check to see if the current bucket exists
+			if err := driver.CheckBucketExists(cr); err != nil {
```
why aren't we attempting to create it if it doesn't exist?

coreydaley replied:
updated in next push

bparees commented on pkg/resource/generator.go:
```diff
@@ -86,8 +96,132 @@ func (g *Generator) list(cr *regopapi.ImageRegistry) ([]Mutator, error) {
+// 2.)  to see if the storage medium name changed and we need to:
+//      a.) check to make sure that we can access the storage or
+//      b.) see if we need to try to create the new storage
+func (g *Generator) syncStorage(cr *regopapi.ImageRegistry, modified *bool) error {
```
We should only be managing the storage in one place, so the fact that you're doing this in both CompleteConfiguration and syncStorage is wrong to me.

Eventually consistent behavior would say the CompleteConfiguration itself should not be responsible for creating/managing the storage.  It can setup whatever initial configuration is needed, but then the standard sync/watch loop should see the configuration (either as created by CompleteConfiguration, or edited/provided by the user) and make the necessary storage checks/updates.

As soon as you start doing it from two different code paths you're creating assumptions about system state and initialization order.

coreydaley replied:
Updated in next push

coreydaley commented:
@bparees ptal
I refactored a bunch of stuff and addressed your comments.
I also merged in my other PR since I needed a few things from it in this one.

I am still working on the test, but I believe everything else is ready for another look.

bparees commented on pkg/storage/s3/s3.go:
```diff
@@ -76,38 +103,127 @@ func (d *driver) Volumes() ([]corev1.Volume, []corev1.VolumeMount, error) {
+					if cr.Spec.Storage.S3.Bucket != "" {
+						break
+					}
+					continue
```
potential infinite loops scare me.  Put a cap of 5000 on this or something and update the status/condition to indicate we failed to generate a unique bucket name.  They can always restart the operator if they want another 5000 runs at it.

coreydaley replied:
updated in next push

bparees commented on pkg/storage/s3/s3.go:
```diff
@@ -129,86 +253,68 @@ func (d *driver) createAndTagBucket(svc *s3.S3, installConfig *installer.Install
-		return err
+		if aerr, ok := err.(awserr.Error); ok {
+			// If we can't set default encryption on the bucket don't return an error
+			// just set the StorageEncrypted condition to false
```
what can the user/admin do to get encryption added to the bucket later?  Do/can they just go encrypt the bucket manually?  Does anything else need to be updated if they do that (e.g. on the registry configuration?).

Is/should there be a way for them to force the operator to attempt to encrypt the bucket again once they've resolved the issue?  (my inclination is to say yes, eventually, but probably not for 4.0).

coreydaley replied:
The user could log into the aws console and turn on default encryption if they want to, and nothing in the registry config needs to be changed.

We could add an option to the config that says the storage should be encrypted, and we check to see if it's encrypted on sync and if it's not, we attempt to encrypt it, if the condition is false and shouldEncrypt is true or something.

coreydaley commented on pkg/resource/generator.go:
```diff
@@ -86,8 +94,140 @@ func (g *Generator) list(cr *regopapi.ImageRegistry) ([]Mutator, error) {
+	if err != nil {
+		return err
+	}
+	if name == "" {
```
updated in next push

coreydaley commented on pkg/resource/generator.go:
```diff
@@ -86,8 +94,140 @@ func (g *Generator) list(cr *regopapi.ImageRegistry) ([]Mutator, error) {
+		}
+	}
+
+	exists, err := driver.StorageExists(cr)
```
updated in next push

coreydaley commented on pkg/resource/generator.go:
```diff
@@ -168,6 +308,16 @@ func (g *Generator) Apply(cr *regopapi.ImageRegistry, modified *bool) error {
 		return fmt.Errorf("unable to remove obsolete routes: %s", err)
 	}
 
+	err = g.syncSecrets(cr, modified)
```
updated in next push

coreydaley commented on pkg/storage/s3/s3.go:
```diff
@@ -10,23 +10,50 @@ import (
 	Config    *opapi.ImageRegistryConfigStorageS3
 }
 
+func (d *driver) getSVC() (*s3.S3, error) {
```
updated in next push

coreydaley commented on pkg/storage/s3/s3.go:
```diff
@@ -76,38 +103,127 @@ func (d *driver) Volumes() ([]corev1.Volume, []corev1.VolumeMount, error) {
+	if cr.Status.Storage.State.S3 == nil {
+		storageChanged = true
+	} else {
+		if cr.Status.Storage.State.S3.Bucket != cr.Spec.Storage.S3.Bucket {
```
updated in next push

coreydaley commented on pkg/storage/s3/s3.go:
```diff
@@ -76,38 +103,127 @@ func (d *driver) Volumes() ([]corev1.Volume, []corev1.VolumeMount, error) {
+			}
+		} else {
+			break
+		}
```
updated in next push

coreydaley commented on pkg/operator/finalizer.go:
```diff
@@ -49,7 +50,16 @@ func (c *Controller) finalizeResources(o *regopapi.ImageRegistry) error {
+
+	err = c.RemoveResources(o)
 	if err != nil {
 		errOp := c.clusterStatus.Update(osapi.OperatorFailing, osapi.ConditionTrue, "unable to remove registry")
```
updated in next push

coreydaley commented on pkg/resource/generator.go:
```diff
@@ -86,8 +94,140 @@ func (g *Generator) list(cr *regopapi.ImageRegistry) ([]Mutator, error) {
+	}
+
+	// If the Image Registry is currently using S3 storage
+	if cr.Spec.Storage.S3 != nil {
```
updated in next push

bparees commented on test/e2e/basic_empty_dir_test.go:
```diff
@@ -1,73 +1,73 @@
+// 	if badlogs {
+// 		testframework.DumpPodLogs(t, logs)
+// 	}
+// }
```
why is this entire test being disabled?

coreydaley replied:
Doh, didin't mean to commit that, i was just trying to run my test by itself so it was faster locally. fixed in next push

bparees commented on test/e2e/unmanaged_test.go:
```diff
@@ -1,80 +1,80 @@
+// 		testframework.DumpOperatorLogs(t, client)
+// 		t.Fatal(err)
+// 	}
+// }
```
and this one?

coreydaley replied:
same, didn't mean to commit

bparees commented on pkg/resource/generator.go:
```diff
@@ -86,8 +93,125 @@ func (g *Generator) list(cr *regopapi.ImageRegistry) ([]Mutator, error) {
+		// Update the deployment template with an annotation of the checksum of the secret
+		// to trigger the Image Registry to be redeployed
+		deployment.Spec.Template.Annotations[fmt.Sprintf("%s-checksum", upSec.ObjectMeta.Name)] = upSecChecksum
+		if _, err := appsclient.Deployments(g.params.Deployment.Namespace).Update(deployment); err != nil {
```
why isn't this done as part of the deployment generator/template?  Having two writers updating the deployment resource is liable to be problematic.

coreydaley replied:
Updated in current push

coreydaley commented:
@bparees ptal
This should be ready for another look.  I separated the aws-sdk-go bump into it's own commit for ease of review.

coreydaley commented:
@bparees bump

bparees commented on pkg/resource/podtemplatespec.go:
```diff
@@ -111,6 +111,7 @@ func makePodTemplateSpec(coreClient coreset.CoreV1Interface, params *parameters.
 	}
 
 	deps := newDependencies()
+	deps.AddSecret(v1alpha1.ImageRegistryPrivateConfigurationUser)
```
shouldn't this be `v1alpha1.ImageRegistryPrivateConfiguration` ?

coreydaley replied:
No, it's already watching that ImageRegistryPrivateConfiguration secret

coreydaley replied:
Fixed in current push

coreydaley commented:
/retest

bparees commented:
/lgtm
